### PR TITLE
local caching

### PR DESCRIPTION
**Pull request by edengal**

### CODE REVIEW

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/TaskIndexer.java [hunk 1]
```diff
@@ -96,34 +101,139 @@ private void handleTimbermillEvents(String env, ZonedDateTime taskIndexerStartTi
...
+                tasksMap.put(adoptedId, adoptedTask);
+            }
+        }
+    }
+
+    private void resolveOrphan(String parentId, Task parentIndexedTask, Map<String, LocalTask> adoptedTasksMap) {
+        if (parentIndexedTask.isOrphan() == null || !parentIndexedTask.isOrphan()){
+            List<String> orphans = orphansCache.getIfPresent(parentId);
+            if (orphans != null) {
+                orphansCache.invalidate(parentId);
```
[the remark this answers is not in the document]
Comment (edengal) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/TaskIndexer.java:163]: To both go to the list

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/TaskIndexer.java [hunk 2]
```diff
@@ -96,34 +101,139 @@ private void handleTimbermillEvents(String env, ZonedDateTime taskIndexerStartTi
...
+            if (tasksMap.containsKey(adoptedId)){
+                tasksMap.get(adoptedId).mergeTask(adoptedTask, adoptedId);
+            }
+            else{
+                tasksMap.put(adoptedId, adoptedTask);
+            }
+        }
+    }
+
+    private void resolveOrphan(String parentId, Task parentIndexedTask, Map<String, LocalTask> adoptedTasksMap) {
```
[the remark this answers is not in the document]
Comment (edengal) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/TaskIndexer.java:159]: 1. TaskIndexer runs in a single thread
2. Done, keep in mind that the amount we "try" to resolve is simply the orphan cache size, which is sent to kamon from a different place.

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/TaskIndexer.java [hunk 3]
```diff
@@ -96,34 +101,139 @@ private void handleTimbermillEvents(String env, ZonedDateTime taskIndexerStartTi
...
+                    LocalTask adoptedTask = tasksCache.getIfPresent(orphanId);
+                    if (adoptedTask == null){
+                        LOG.warn("Missing task from local cache {}", orphanId);
+                    }
+                    else {
+                        adoptedTask.setOrphan(false);
+
+                        populateParentParams(adoptedTask, parentIndexedTask);
+                        adoptedTasksMap.put(orphanId, adoptedTask);
+                        resolveOrphan(orphanId, adoptedTask, adoptedTasksMap);
```
[the remark this answers is not in the document]
Comment (edengal) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/TaskIndexer.java:174]: Good point, the chances are slim, but I will add a security measure

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/IndexRetryManager.java [hunk 4]
```diff
@@ -35,53 +34,49 @@ public IndexRetryManager(int numOfElasticSearchActionsTries, int maxBulkIndexFet
...
 				}
-				else{
-					// SUCCESS
-					LOG.info(FLOW_ID_LOG + " Bulk #{} Try # {} finished successfully. Took: {} millis.", flowId, bulkNum, tryNum, retryResponse.getTook().millis());
-					if (dbBulkRequest.getTimesFetched() > 0 ){
-						KamonConstants.TASKS_FETCHED_FROM_DISK_HISTOGRAM.withTag("outcome","success").record(1);
+				LOG.debug(FLOW_ID_LOG + " Bulk #{} Batch of {} index requests sent to Elasticsearch. Batch size: {} bytes", flowId, bulkNum, dbBulkRequest.numOfActions(), dbBulkRequest.estimatedSize());
+				BulkResponse response = bulker.bulk(dbBulkRequest);
+					resList.add(response);
+					if (!response.hasFailures()) {
```
[the remark this answers is not in the document]
Comment (edengal) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/IndexRetryManager.java:49]: not sure which db handling you mean

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/TaskIndexer.java [hunk 5]
```diff
@@ -96,34 +101,139 @@ private void handleTimbermillEvents(String env, ZonedDateTime taskIndexerStartTi
...
+                cachedTask.mergeTask(localTask, id);
+            }
+        }
+    }
+
+    private Map<String, Task> getMissingParents(Set<String> parentIds, String flowId) {
+        
+        int missingParentAmount = parentIds.size();
+        KamonConstants.MISSING_PARENTS_HISTOGRAM.withoutTags().record(missingParentAmount);
+        LOG.info(FLOW_ID_LOG + " Fetching {} missing parents", flowId, missingParentAmount);
```
[the remark this answers is not in the document]
Comment (edengal) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/TaskIndexer.java:218]: NICE! added.

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 6]
```diff
@@ -400,27 +336,39 @@ BulkResponse bulk(DbBulkRequest request) throws IOException {
...
-				overallFailedRequests += failedRequests;
+        Map<String, String> overallIdToIndex = Maps.newHashMap();
+		for (Future<List<BulkResponse>> futureRequest : futuresRequests) {
+            try {
+				List<BulkResponse> bulkResponses = futureRequest.get();
+				Map<String, String> idToIndexMap = getIdToIndexMap(bulkResponses);
+				overallIdToIndex.putAll(idToIndexMap);
 			} catch (InterruptedException e) {
 				LOG.error(FLOW_ID_LOG + " Bulk #{} An error was thrown while indexing a batch, going to retry", flowId, bulkNum, e);
-				sendDbBulkRequest(futureRequest.getRight(), flowId, bulkNum);
```
[the remark this answers is not in the document]
Comment (edengal) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:414]: Just not in InterruptedException case

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 7]
```diff
@@ -545,40 +493,43 @@ private BoolQueryBuilder getLatestPartialsQuery() {
...
 		if (!tasksToMigrateIntoNewIndex.isEmpty()) {
 			LOG.info(FLOW_ID_LOG + " Migrating {} tasks to new index [{}]", flowId, tasksToMigrateIntoNewIndex.size(), currentIndex);
-			int failedRequests = index(tasksToMigrateIntoNewIndex, currentIndex, flowId);
-			if (failedRequests > 0){
-				LOG.info(FLOW_ID_LOG + " There were {} failed migration requests", flowId, failedRequests);
-				KamonConstants.PARTIAL_TASKS_FAILED_TO_MIGRATED_HISTOGRAM.withoutTags().record(failedRequests);
-			}
+			index(tasksToMigrateIntoNewIndex, currentIndex, flowId);
+
+			//TODO fix kamon
```
[the remark this answers is not in the document]
Comment (edengal) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:498]: Pushed

File: timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/IndexRetryManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (DvirKatz) [reply timbermill-java/timbermill-local/src/main/java/com/datorama/oss/timbermill/common/disk/IndexRetryManager.java:47]: db / failure handling  -- all the code inside the "if"

**Pull request merged**